File: hand_tracking.py
"""
hand_tracking.py
----------------
Lógica de detección/encuadre de manos con MediaPipe, suavizado de
coordenadas, rotación del visor y dibujado de bordes.
"""

import cv2
import numpy as np
import random

from config import MIN_BOX_SIZE, BOX_PADDING, SMOOTHING


# ----------------------------------------------------------------------------
# UTILIDADES DE SUAVIZADO
# ----------------------------------------------------------------------------
class SmoothBox:
    """Suaviza las coordenadas del cuadro para evitar temblores frame a frame."""

    def __init__(self, alpha=SMOOTHING):
        self.alpha = alpha
        self.box = None  # (x1, y1, x2, y2)

    def update(self, new_box):
        if self.box is None:
            self.box = list(new_box)
        else:
            for i in range(4):
                self.box[i] = self.alpha * self.box[i] + (1 - self.alpha) * new_box[i]
        return tuple(int(v) for v in self.box)


class SmoothRotBox:
    """Suaviza (cx, cy, w, h, angle) para evitar temblores/saltos frame a
    frame, con manejo especial del angulo para que no 'salte' al cruzar
    +-90 grados."""

    def __init__(self, alpha=SMOOTHING):
        self.alpha = alpha
        self.state = None  # [cx, cy, w, h, angle]

    def update(self, new_state):
        cx, cy, w_, h_, angle = new_state
        if self.state is None:
            self.state = [cx, cy, w_, h_, angle]
        else:
            a = self.alpha
            # angulo: interpolar por el camino mas corto
            prev_angle = self.state[4]
            diff = angle - prev_angle
            if diff > 90:
                angle -= 180
            elif diff < -90:
                angle += 180
            self.state[0] = a * self.state[0] + (1 - a) * cx
            self.state[1] = a * self.state[1] + (1 - a) * cy
            self.state[2] = a * self.state[2] + (1 - a) * w_
            self.state[3] = a * self.state[3] + (1 - a) * h_
            self.state[4] = a * prev_angle + (1 - a) * angle
        return tuple(self.state)


# ----------------------------------------------------------------------------
# LÓGICA DE ENCUADRE SEGÚN MANOS
# ----------------------------------------------------------------------------
def landmarks_to_px(hand_landmarks, w, h):
    return np.array([(int(lm.x * w), int(lm.y * h)) for lm in hand_landmarks.landmark])


def box_from_two_hands(hands_px, w, h):
    """Bounding box que envuelve todos los puntos de ambas manos."""
    all_pts = np.vstack(hands_px)
    x1, y1 = all_pts[:, 0].min(), all_pts[:, 1].min()
    x2, y2 = all_pts[:, 0].max(), all_pts[:, 1].max()
    x1 -= BOX_PADDING
    y1 -= BOX_PADDING
    x2 += BOX_PADDING
    y2 += BOX_PADDING
    return clamp_box(x1, y1, x2, y2, w, h)


def box_from_finger_frame(hands_px_list, w, h):
    """Usa pulgar (4) e índice (8) de cada mano para armar un encuadre tipo
    'visor de cámara'. Con una sola mano usa esos dos dedos; con dos manos,
    toma el punto extremo de cada mano para ampliar el marco."""
    pts = []
    for hp in hands_px_list:
        pts.append(hp[4])   # punta del pulgar
        pts.append(hp[8])   # punta del índice
    pts = np.array(pts)
    x1, y1 = pts[:, 0].min(), pts[:, 1].min()
    x2, y2 = pts[:, 0].max(), pts[:, 1].max()
    x1 -= BOX_PADDING // 2
    y1 -= BOX_PADDING // 2
    x2 += BOX_PADDING // 2
    y2 += BOX_PADDING // 2
    return clamp_box(x1, y1, x2, y2, w, h)


def rotated_rect_from_hands(hands_px_list, w, h):
    """Calcula un rectangulo ROTADO (centro, tamaño, angulo) a partir de las
    manos, para el look tipo 'visor inclinado' de la referencia:
      - Con DOS manos: el angulo es el que forma la linea entre ambas manos
        (usando la muñeca, landmark 0) respecto a la horizontal. El rectangulo
        se arma con las puntas de pulgar+indice de ambas manos, luego se
        proyecta sobre los ejes rotados.
      - Con UNA mano: el angulo es el que forma el vector pulgar->indice.
    Retorna (cx, cy, rect_w, rect_h, angle_deg).
    """
    pts_frame = []
    for hp in hands_px_list:
        pts_frame.append(hp[4])
        pts_frame.append(hp[8])
    pts_frame = np.array(pts_frame, dtype=np.float32)

    if len(hands_px_list) >= 2:
        c1, c2 = hands_px_list[0][0], hands_px_list[1][0]
        vec = c2.astype(np.float32) - c1.astype(np.float32)
    else:
        thumb, index = hands_px_list[0][4], hands_px_list[0][8]
        vec = index.astype(np.float32) - thumb.astype(np.float32)

    angle = float(np.degrees(np.arctan2(vec[1], vec[0])))
    # normalizar para que el rectangulo no quede "de cabeza" (evita saltos
    # de +-180 que harian girar el marco bruscamente)
    if angle > 90:
        angle -= 180
    elif angle < -90:
        angle += 180

    cx, cy = pts_frame[:, 0].mean(), pts_frame[:, 1].mean()

    # proyectar los puntos sobre los ejes rotados para medir ancho/alto reales
    theta = np.radians(angle)
    cos_t, sin_t = np.cos(theta), np.sin(theta)
    rel = pts_frame - np.array([cx, cy], dtype=np.float32)
    local_x = rel[:, 0] * cos_t + rel[:, 1] * sin_t
    local_y = -rel[:, 0] * sin_t + rel[:, 1] * cos_t

    rect_w = float(local_x.max() - local_x.min()) + BOX_PADDING
    rect_h = float(local_y.max() - local_y.min()) + BOX_PADDING
    rect_w = max(MIN_BOX_SIZE, rect_w)
    rect_h = max(MIN_BOX_SIZE, rect_h)

    return cx, cy, rect_w, rect_h, angle


def get_rotated_patch(frame, cx, cy, rect_w, rect_h, angle_deg):
    """Extrae de 'frame' el contenido dentro del rectangulo rotado, ya
    enderezado (como si se hiciera un crop normal), listo para aplicarle
    el efecto de imagen."""
    rect_w_i, rect_h_i = max(2, int(rect_w)), max(2, int(rect_h))
    M = cv2.getRotationMatrix2D((cx, cy), angle_deg, 1.0)
    h, w = frame.shape[:2]
    rotated_full = cv2.warpAffine(frame, M, (w, h), flags=cv2.INTER_LINEAR,
                                   borderMode=cv2.BORDER_REFLECT)
    x1 = int(cx - rect_w_i / 2)
    y1 = int(cy - rect_h_i / 2)
    x2, y2 = x1 + rect_w_i, y1 + rect_h_i
    x1c, y1c = max(0, x1), max(0, y1)
    x2c, y2c = min(w, x2), min(h, y2)
    if x2c <= x1c or y2c <= y1c:
        return None
    patch = rotated_full[y1c:y2c, x1c:x2c]
    # si el rectangulo ideal se salia del frame, el patch queda mas chico
    # que rect_w_i x rect_h_i; se reescala para mantener tamaño consistente
    if patch.shape[0] != rect_h_i or patch.shape[1] != rect_w_i:
        patch = cv2.resize(patch, (rect_w_i, rect_h_i))
    return patch


def paste_rotated_patch(display, patch, cx, cy, rect_w, rect_h, angle_deg):
    """Pega 'patch' (ya procesado, alineado horizontalmente) de vuelta en
    'display' rotandolo angle_deg grados alrededor de (cx, cy), con una
    mascara para que solo se pinte el area del rectangulo rotado (encima de
    lo que ya habia, como un sello). Retorna (imagen_resultado, mascara_rotada)."""
    h, w = display.shape[:2]
    rect_w_i, rect_h_i = max(2, int(rect_w)), max(2, int(rect_h))

    ph, pw = patch.shape[:2]
    if ph != rect_h_i or pw != rect_w_i:
        patch = cv2.resize(patch, (rect_w_i, rect_h_i))

    # Lienzo del tamaño del frame, con el patch ya colocado (sin rotar) en
    # su posicion centrada en (cx, cy)
    canvas = np.zeros_like(display)
    mask = np.zeros((h, w), dtype=np.uint8)
    x1 = int(cx - rect_w_i / 2)
    y1 = int(cy - rect_h_i / 2)
    x2, y2 = x1 + rect_w_i, y1 + rect_h_i
    x1c, y1c = max(0, x1), max(0, y1)
    x2c, y2c = min(w, x2), min(h, y2)
    if x2c <= x1c or y2c <= y1c:
        empty_mask = np.zeros((h, w), dtype=np.uint8)
        return display, empty_mask

    px1, py1 = x1c - x1, y1c - y1
    px2, py2 = px1 + (x2c - x1c), py1 + (y2c - y1c)
    canvas[y1c:y2c, x1c:x2c] = patch[py1:py2, px1:px2]
    mask[y1c:y2c, x1c:x2c] = 255

    # Rotar canvas y mascara alrededor de (cx, cy) por -angle_deg para
    # "devolver" la inclinacion original a la escena
    M_inv = cv2.getRotationMatrix2D((cx, cy), -angle_deg, 1.0)
    rotated_canvas = cv2.warpAffine(canvas, M_inv, (w, h), flags=cv2.INTER_LINEAR)
    rotated_mask = cv2.warpAffine(mask, M_inv, (w, h), flags=cv2.INTER_LINEAR)

    mask_3ch = cv2.merge([rotated_mask] * 3).astype(np.float32) / 255.0
    result = display.astype(np.float32) * (1 - mask_3ch) + rotated_canvas.astype(np.float32) * mask_3ch
    return result.astype(np.uint8), rotated_mask


def draw_broken_glass_border(display, rotated_mask, color=(220, 60, 230), thickness=3,
                              jitter=6, n_extra_lines=5):
    """Dibuja un borde tipo 'vidrio roto' (lineas irregulares, no un
    rectangulo perfecto) siguiendo el contorno de la mascara rotada, en un
    color magenta brillante como en la referencia visual."""
    contours, _ = cv2.findContours(rotated_mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    if not contours:
        return display
    contour = max(contours, key=cv2.contourArea)

    # glow: primero una linea gruesa semi-transparente para dar brillo,
    # luego el borde nitido encima
    overlay = display.copy()
    cv2.drawContours(overlay, [contour], -1, color, thickness + 5)
    display = cv2.addWeighted(overlay, 0.35, display, 0.65, 0)

    jittered = contour.copy().astype(np.int32)
    if len(jittered) > 2:
        noise = np.random.randint(-jitter, jitter + 1, jittered.shape)
        jittered = jittered + noise
    cv2.drawContours(display, [jittered], -1, color, thickness)

    # unas "grietas" extra: lineas cortas aleatorias que salen del contorno,
    # como fracturas de vidrio
    pts = contour.reshape(-1, 2)
    if len(pts) > 5:
        for _ in range(n_extra_lines):
            idx = random.randint(0, len(pts) - 1)
            p1 = tuple(pts[idx])
            crack_len = random.randint(10, 25)
            crack_angle = random.uniform(0, 2 * np.pi)
            p2 = (int(p1[0] + crack_len * np.cos(crack_angle)),
                  int(p1[1] + crack_len * np.sin(crack_angle)))
            cv2.line(display, p1, p2, color, 1, cv2.LINE_AA)

    return display


def box_from_center_zoom(hands_px_list, w, h):
    """Centra el cuadro en el punto medio de las manos (usando la muñeca,
    landmark 0) y escala el tamaño según la distancia entre ellas."""
    centers = [hp[0] for hp in hands_px_list]
    if len(centers) == 1:
        cx, cy = centers[0]
        size = 200
    else:
        c1, c2 = centers[0], centers[1]
        cx, cy = (c1 + c2) // 2
        size = int(np.linalg.norm(c1 - c2))
        size = max(MIN_BOX_SIZE * 2, size)
    x1, y1 = cx - size // 2, cy - size // 2
    x2, y2 = cx + size // 2, cy + size // 2
    return clamp_box(x1, y1, x2, y2, w, h)


def clamp_box(x1, y1, x2, y2, w, h):
    x1 = max(0, min(int(x1), w - 2))
    y1 = max(0, min(int(y1), h - 2))
    x2 = max(x1 + MIN_BOX_SIZE, min(int(x2), w))
    y2 = max(y1 + MIN_BOX_SIZE, min(int(y2), h))
    return (x1, y1, x2, y2)


MODES = ["dos_manos_bbox", "encuadre_dedos", "zoom_centro"]
MODE_LABELS = {
    "dos_manos_bbox": "Cuadro: bounding box de ambas manos",
    "encuadre_dedos": "Cuadro: pulgar+indice (visor de camara)",
    "zoom_centro": "Cuadro: centro entre manos + zoom por distancia",
}


def compute_box(mode, hands_px_list, w, h):
    if mode == "dos_manos_bbox":
        return box_from_two_hands(hands_px_list, w, h)
    elif mode == "encuadre_dedos":
        return box_from_finger_frame(hands_px_list, w, h)
    elif mode == "zoom_centro":
        return box_from_center_zoom(hands_px_list, w, h)
    return box_from_two_hands(hands_px_list, w, h)

File: Test/hand_tracking.js
/**
 * Test/hand_tracking.js
 * ---------------------
 * Lógica de tracking de manos, suavizado de coordenadas, encuadre
 * trigonométrico y dibujo del borde "vidrio roto" estilo Visor Roto.
 */

// Constantes de tracking
const MIN_BOX_SIZE = 40;
const BOX_PADDING = 25;
const SMOOTHING = 0.35;

export const MODES = ['dos_manos_bbox', 'encuadre_dedos', 'zoom_centro'];
export const MODE_LABELS = {
    dos_manos_bbox: 'Dos manos',
    encuadre_dedos: 'Visor de dedos',
    zoom_centro: 'Zoom centro'
};

/**
 * Suaviza las coordenadas del cuadro (x1, y1, x2, y2) para evitar temblores.
 */
export class SmoothBox {
    constructor(alpha = SMOOTHING) {
        this.alpha = alpha;
        this.box = null;
    }

    update(newBox) {
        if (!newBox) return this.box;
        if (!this.box) {
            this.box = [...newBox];
        } else {
            for (let i = 0; i < 4; i++) {
                this.box[i] = this.alpha * this.box[i] + (1 - this.alpha) * newBox[i];
            }
        }
        return this.box.map(v => Math.round(v));
    }
}

/**
 * Suaviza (cx, cy, w, h, angle) con manejo especial de cruces +-90 grados.
 */
export class SmoothRotBox {
    constructor(alpha = SMOOTHING) {
        this.alpha = alpha;
        this.state = null; // [cx, cy, w, h, angle]
    }

    update(newState) {
        if (!newState) return this.state;
        let [cx, cy, w, h, angle] = newState;

        if (!this.state) {
            this.state = [cx, cy, w, h, angle];
        } else {
            const a = this.alpha;
            let prevAngle = this.state[4];
            let diff = angle - prevAngle;
            if (diff > 90) angle -= 180;
            else if (diff < -90) angle += 180;

            this.state[0] = a * this.state[0] + (1 - a) * cx;
            this.state[1] = a * this.state[1] + (1 - a) * cy;
            this.state[2] = a * this.state[2] + (1 - a) * w;
            this.state[3] = a * this.state[3] + (1 - a) * h;
            this.state[4] = a * prevAngle + (1 - a) * angle;
        }
        return [...this.state];
    }
}

export function landmarksToPx(landmarks, w, h) {
    return landmarks.map(lm => ({
        x: Math.round(lm.x * w),
        y: Math.round(lm.y * h)
    }));
}

function clampBox(x1, y1, x2, y2, w, h) {
    x1 = Math.max(0, Math.min(w - 1, x1));
    y1 = Math.max(0, Math.min(h - 1, y1));
    x2 = Math.max(0, Math.min(w - 1, x2));
    y2 = Math.max(0, Math.min(h - 1, y2));
    if (x2 - x1 < MIN_BOX_SIZE) {
        const cx = Math.floor((x1 + x2) / 2);
        x1 = Math.max(0, cx - Math.floor(MIN_BOX_SIZE / 2));
        x2 = Math.min(w - 1, x1 + MIN_BOX_SIZE);
    }
    if (y2 - y1 < MIN_BOX_SIZE) {
        const cy = Math.floor((y1 + y2) / 2);
        y1 = Math.max(0, cy - Math.floor(MIN_BOX_SIZE / 2));
        y2 = Math.min(h - 1, y1 + MIN_BOX_SIZE);
    }
    return [x1, y1, x2, y2];
}

export function boxFromTwoHands(handsPx, w, h) {
    const allPts = handsPx.flat();
    if (!allPts.length) return null;
    let x1 = Infinity, y1 = Infinity, x2 = -Infinity, y2 = -Infinity;
    for (const pt of allPts) {
        if (pt.x < x1) x1 = pt.x;
        if (pt.y < y1) y1 = pt.y;
        if (pt.x > x2) x2 = pt.x;
        if (pt.y > y2) y2 = pt.y;
    }
    return clampBox(x1 - BOX_PADDING, y1 - BOX_PADDING, x2 + BOX_PADDING, y2 + BOX_PADDING, w, h);
}

export function boxFromFingerFrame(handsPxList, w, h) {
    const pts = [];
    for (const hp of handsPxList) {
        if (hp[4]) pts.push(hp[4]); // pulgar
        if (hp[8]) pts.push(hp[8]); // índice
    }
    if (!pts.length) return null;
    let x1 = Infinity, y1 = Infinity, x2 = -Infinity, y2 = -Infinity;
    for (const pt of pts) {
        if (pt.x < x1) x1 = pt.x;
        if (pt.y < y1) y1 = pt.y;
        if (pt.x > x2) x2 = pt.x;
        if (pt.y > y2) y2 = pt.y;
    }
    const pad = Math.floor(BOX_PADDING / 2);
    return clampBox(x1 - pad, y1 - pad, x2 + pad, y2 + pad, w, h);
}

export function zoomCenterBox(handsPxList, w, h) {
    const baseBox = boxFromFingerFrame(handsPxList, w, h) || boxFromTwoHands(handsPxList, w, h);
    if (!baseBox) return null;
    const [x1, y1, x2, y2] = baseBox;
    const cx = (x1 + x2) / 2;
    const cy = (y1 + y2) / 2;
    const bw = (x2 - x1) * 0.7;
    const bh = (y2 - y1) * 0.7;
    return clampBox(Math.round(cx - bw / 2), Math.round(cy - bh / 2), Math.round(cx + bw / 2), Math.round(cy + bh / 2), w, h);
}

export function computeBox(mode, handsPx, w, h) {
    if (mode === 'dos_manos_bbox') return boxFromTwoHands(handsPx, w, h);
    if (mode === 'encuadre_dedos') return boxFromFingerFrame(handsPx, w, h);
    if (mode === 'zoom_centro') return zoomCenterBox(handsPx, w, h);
    return boxFromFingerFrame(handsPx, w, h);
}

export function rotatedRectFromHands(handsPxList, w, h) {
    const pts = [];
    for (const hp of handsPxList) {
        if (hp[4]) pts.push(hp[4]);
        if (hp[8]) pts.push(hp[8]);
    }
    if (!pts.length) return null;

    let vecX, vecY;
    if (handsPxList.length >= 2) {
        const c1 = handsPxList[0][0]; // muñeca
        const c2 = handsPxList[1][0];
        vecX = c2.x - c1.x;
        vecY = c2.y - c1.y;
    } else {
        const thumb = handsPxList[0][4];
        const index = handsPxList[0][8];
        vecX = index.x - thumb.x;
        vecY = index.y - thumb.y;
    }

    let angle = Math.atan2(vecY, vecX) * (180 / Math.PI);
    if (angle > 90) angle -= 180;
    else if (angle < -90) angle += 180;

    let sumX = 0, sumY = 0;
    for (const p of pts) {
        sumX += p.x;
        sumY += p.y;
    }
    const cx = sumX / pts.length;
    const cy = sumY / pts.length;

    const rad = angle * (Math.PI / 180);
    const cosT = Math.cos(rad);
    const sinT = Math.sin(rad);

    let minLX = Infinity, maxLX = -Infinity;
    let minLY = Infinity, maxLY = -Infinity;

    for (const p of pts) {
        const rx = p.x - cx;
        const ry = p.y - cy;
        const lx = rx * cosT + ry * sinT;
        const ly = -rx * sinT + ry * cosT;
        if (lx < minLX) minLX = lx;
        if (lx > maxLX) maxLX = lx;
        if (ly < minLY) minLY = ly;
        if (ly > maxLY) maxLY = ly;
    }

    let rectW = Math.max(MIN_BOX_SIZE, (maxLX - minLX) + BOX_PADDING);
    let rectH = Math.max(MIN_BOX_SIZE, (maxLY - minLY) + BOX_PADDING);

    return [cx, cy, rectW, rectH, angle];
}

/**
 * Dibuja un marco estilo vidrio roto cyberpunk con brillo magenta.
 */
export function drawBrokenGlassBorder(ctx, cx, cy, w, h, angleDeg) {
    ctx.save();
    ctx.translate(cx, cy);
    ctx.rotate(angleDeg * Math.PI / 180);

    const hw = w / 2;
    const hh = h / 2;

    ctx.strokeStyle = '#c83ce6';
    ctx.lineWidth = 3;
    ctx.shadowColor = '#c83ce6';
    ctx.shadowBlur = 12;

    // Trazado irregular tipo fractura
    ctx.beginPath();
    const jitter = 5;
    ctx.moveTo(-hw, -hh);
    ctx.lineTo(0 + (Math.random() - 0.5) * jitter, -hh + (Math.random() - 0.5) * jitter);
    ctx.lineTo(hw, -hh);
    ctx.lineTo(hw + (Math.random() - 0.5) * jitter, 0 + (Math.random() - 0.5) * jitter);
    ctx.lineTo(hw, hh);
    ctx.lineTo(0 + (Math.random() - 0.5) * jitter, hh + (Math.random() - 0.5) * jitter);
    ctx.lineTo(-hw, hh);
    ctx.lineTo(-hw + (Math.random() - 0.5) * jitter, 0 + (Math.random() - 0.5) * jitter);
    ctx.closePath();
    ctx.stroke();

    // Grietas sutiles que salen de las esquinas
    ctx.lineWidth = 1.5;
    ctx.beginPath();
    ctx.moveTo(-hw, -hh);
    ctx.lineTo(-hw + 14, -hh + 14);
    ctx.moveTo(hw, -hh);
    ctx.lineTo(hw - 14, -hh + 14);
    ctx.moveTo(hw, hh);
    ctx.lineTo(hw - 14, hh - 14);
    ctx.moveTo(-hw, hh);
    ctx.lineTo(-hw + 14, hh - 14);
    ctx.stroke();

    ctx.restore();
}

/**
 * Dibuja el marco normal rectangular con esquinas estilo HUD.
 */
export function drawHudBorder(ctx, x1, y1, x2, y2) {
    ctx.save();
    ctx.strokeStyle = '#00e5ff';
    ctx.lineWidth = 2;
    ctx.shadowColor = '#00e5ff';
    ctx.shadowBlur = 8;
    ctx.strokeRect(x1, y1, x2 - x1, y2 - y1);

    // Esquinas acentuadas
    ctx.lineWidth = 3;
    const corner = 16;
    ctx.beginPath();
    // Sup Izq
    ctx.moveTo(x1, y1 + corner); ctx.lineTo(x1, y1); ctx.lineTo(x1 + corner, y1);
    // Sup Der
    ctx.moveTo(x2 - corner, y1); ctx.lineTo(x2, y1); ctx.lineTo(x2, y1 + corner);
    // Inf Der
    ctx.moveTo(x2, y2 - corner); ctx.lineTo(x2, y2); ctx.lineTo(x2 - corner, y2);
    // Inf Izq
    ctx.moveTo(x1 + corner, y2); ctx.lineTo(x1, y2); ctx.lineTo(x1, y2 - corner);
    ctx.stroke();
    ctx.restore();
}

/**
 * Dibuja el esqueleto de las manos (landmarks y conexiones).
 */
export function drawHandSkeleton(ctx, landmarksList, w, h) {
    const CONNECTIONS = [
        [0, 1], [1, 2], [2, 3], [3, 4],       // pulgar
        [0, 5], [5, 6], [6, 7], [7, 8],       // índice
        [5, 9], [9, 10], [10, 11], [11, 12],  // medio
        [9, 13], [13, 14], [14, 15], [15, 16],// anular
        [13, 17], [17, 18], [18, 19], [19, 20],// meñique
        [0, 17]                               // palma
    ];

    ctx.save();
    for (const lm of landmarksList) {
        const pts = lm.map(p => ({ x: p.x * w, y: p.y * h }));

        // Conexiones
        ctx.strokeStyle = 'rgba(54, 214, 231, 0.6)';
        ctx.lineWidth = 2;
        ctx.beginPath();
        for (const [i, j] of CONNECTIONS) {
            ctx.moveTo(pts[i].x, pts[i].y);
            ctx.lineTo(pts[j].x, pts[j].y);
        }
        ctx.stroke();

        // Puntos
        for (let i = 0; i < pts.length; i++) {
            ctx.fillStyle = (i === 4 || i === 8) ? '#ff3366' : '#c83ce6';
            ctx.beginPath();
            ctx.arc(pts[i].x, pts[i].y, (i === 4 || i === 8) ? 4.5 : 3, 0, Math.PI * 2);
            ctx.fill();
        }
    }
    ctx.restore();
}
